### ISSUE TITLE
Ribbon retries

### ISSUE THREAD

**Issue by wbauern** [spring-cloud/spring-cloud-netflix#1784]
Where I can find what specific exceptions will be retried when using RestTemplate configured to use Ribbon retries (ie. spring.cloud.loadbalancer.retry=true)?

Thanks!

**Comment by ryanjbaxter** [comment]
What version of Spring Cloud are you using?  It might be useful to read through issue #1551.

**Comment by wbauern** [comment]
Camden SR5.  I'll take a look at that issue...

**Comment by wbauern** [comment]
Sorry, still didn't help.  I saw that Ribbon retries were was recently refactored to use Spring Retry so I've looked through the Spring Cloud Netflix code looking for `@Retryable` but had no luck.  Can you point me at the package/class that would be defining the Exceptions it is going to retry on?

**Comment by spencergibb** [comment]
In ribbon it is `ConnectException.class, SocketTimeoutException.class`, but I don't know if the new Spring Retry functionality takes exception types into account. @ryanjbaxter?

**Comment by ryanjbaxter** [comment]
With the new retry logic we retry on all exceptions.  I would suggest upgrading to `Camden.SR6`.  Is there a specific scenario where you do not see a retry happening when you expect it to?  Do you have Spring Retry on the classpath of your app?

**Comment by wbauern** [comment]
From what I saw with some testing it looks like the retries are working.  I was checking to see if it was only supposed to be for connection exceptions.  We usually wrap our remote resttemplate calls with hystrix commands and let it dictate what to do on certain failures (ex. fallback hits a different type of service on a 500 response status).  I was hoping to let the resttemplate retry only handle the case where it couldn't connect to a specific instance and then quickly pick another.  Is there a way to customize it to only retry on a set of given exceptions?

Also, we are thinking of moving to Feign clients for some of our services.  Does it have the same behavior?

**Comment by ryanjbaxter** [comment]
Hum, this is not a use case I thought of, however here is something you can try

Try providing a `Bean` of type `LoadBalanceRetryPolicyFactory`.  You might want to extend [`RibbonLoadBalancedRetryPolicyFactory`](https://github.com/spring-cloud/spring-cloud-netflix/blob/fabf90bed4aa090cf956494ab9bf821d41cb3bca/spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancedRetryPolicyFactory.java).   In that class there is are `canRetry` and `registerThrowable` methods that you might be able to use to do what you want.

There is no conditional on the bean we create in `RibbonAutoConfiguration` so you will probably need to add `@Primary` to your bean.
https://github.com/spring-cloud/spring-cloud-netflix/blob/5bf87c4da1e490bdba03891d1e3c0c1bee6992ac/spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonAutoConfiguration.java#L93

> Also, we are thinking of moving to Feign clients for some of our services. Does it have the same behavior?
Yes

Let me know how that works and we can make the necessary changes to make this easier in the future.

**Comment by pfournea** [comment]
Hi @ryanjbaxter ,

I have a workaround for retry policy that takes into account the exclusion of specific exceptions. It work's like this:

* I add the @Retryable annotation on the service (or repository) interface. The implementation of this interfaces uses a resttemplate (with the @Loadbalanced annotation)

`@Retryable(backoff=@Backoff(delay = 100, maxDelay = 1000, multiplier = 2.0, random = true), exclude = {HttpClientErrorException.class, FunctionalValidationException.class})
public interface CountryRepository {
    String getCountryCodeByIso(String isoCode);
}`

* disable the loadbalancer retry mechanism in the resttemplate
`spring.cloud.loadbalancer.retry.enabled=false`

When errors occure the never-retry policy is triggered in the resttemplate, the retry template around the service/repository takes care of retrying based on the configuration passed in the @Retryable annotation.

I think the problem with this solution is that I can't control the service instances that is being used when retrying. If I retry 3 times, there is a change that I use 3 times the same service instance.

Are there better solutions for this than creating an own implementation of a LoadBalancedRetryPolicy(Factory)?

the @Retryable annotation is very convenient and powerfull. Maybe an extend of the @Loadbalanced annotation could be a solution the future (@LoadBalanced(@Retryable(...))?

Thanks for feedback.

Kind regards,

Peter F

**Comment by ryanjbaxter** [comment]
> If I retry 3 times, there is a change that I use 3 times the same service instance.

This shouldnt be the case, the load balancer should be using a round robin policy.  You can provide your own Ribbon load balancer policy if you want as well.

You are also enabling retry on a per api basis in your approach.  I think the general appoach of having it enabled for everything by default and then disabling it on a per service basis via configuration seems to work for most people.

What you are really asking though is if we can provide a way to exclude certain exceptions from being retried, correct?

**Comment by pfournea** [comment]
Correct! An easy way to exclude exceptions and/or specific http response statuses from retry (eg 404 should not result in a retry).

Thanks,

Peter F
